resholve: init at 0.4.0

abathur opened the pull request:
###### Motivation for this change

Give the Nix ecosystem a new superpower: resolving external dependencies in shell scripts.

Shell scripts/libraries with external dependencies are an odd duck in the Nix ecosystem. Packaging them with Nix doesn't ensure dependencies are available at runtime or address the possibility that the script could run executables provided by packages other than those specified as dependencies, for example.

###### More information

This PR packages a project of mine, https://github.com/abathur/resholve, and provides a Nix build function. resholve is built atop the parser from the Oil shell project. A fair portion of the Nix code in this PR is just packaging/building a development version of Oil. 

The PR itself isn't terribly useful for seeing what it _does_; it'll help to look at some additional resources:
- #107182 shows how I've used it to package a new shell library.
- [This commit](https://github.com/abathur/resholve/commit/7288d42e1fa78e2b228884b495d998e6a35d7708) shows (roughly, I'm synthesizing this) how you'd resholve the existing shunit2, and how it would change the script that ends up in the store.
- The repo includes two demos (shell API, Nix API), which you can read about (including invocation instructions) at https://github.com/abathur/resholve/blob/master/demos.md.

FWIW, I dogfood resholve in my own darwin-configuration, where it is responsible for building a chain of packages structured like:
- my bashrc
   - a bash history-management module I'm developing
      - an event-based shell-profile "orchestration" framework I'm developing to support both of the above
         - a third-party bash event/promise library (now PRed in #107182)

<details>
<summary>A question came up during review about whether some of the packages in `deps.nix` should be split out into standalone packages. I've resolved the thread, but others may have the same question so I'm keeping the answer somewhere easier to find:</summary>

> > BTW, having oilshell and maybe oildev available in `all-packages` is probably a good idea, since those packages are kinda interesting. You could, for example, open a PR with only `oildshell`, `oildev` and maybe `py-yajl`, making it easier to review only those dependencies. Afterwards you could make this PR depends on those dependencies.
> 
> I have reservations about this.
> 
> I'm happy to break these out if there's a clear imperative, but they are all pretty idiomatic (and all ultimately subservient to the oildev package itself). I don't anticipate them adding standalone value to nixpkgs in the short run.
> 
> 1. The Oil shell already has a binary user package included here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/shells/oil/default.nix
> 2. oildev isn't (for now) a _real thing_ that the Oil project intentionally publishes--it's a fiction created by forcibly patching-and-packaging Oil's python code in order to build resholve
>    
>    * Because these Python APIs aren't intentionally published for external consumption and Oil is still a fast-changing young project, they aren't stable. I'd be very surprised if more than one consumer could share a single version of it for long.
>    * AFAIK resholve is the only consumer of this. No one's asked me about using it.
>    * resholve still needs [~140 lines of code](https://github.com/abathur/resholve/blob/d990fc06b504d761ff811bd1ad8075bdceb61dd0/resholve#L23-L161) just to monkeypatch all of oil's packages into a unified oil.* namespace. If anyone used it without this shim, all of its subpackages would leak into the root Python package namespace.
>    * I can't guarantee any Oil code paths that resholve doesn't exercise actually work.
>    * Because resholve is very tightly coupled to Oil's internal Python APIs, there's a fair chance any update of this would break resholve.
> 3. [py-yajl](https://github.com/oilshell/py-yajl) is Oil's fork of https://github.com/rtyler/py-yajl. They've modified it, dropped tests they don't need, and cut code they don't use. I have no idea how suitable it is or isn't for general use. Oil has made no attempt to promote their py-yajl fork, it has no stars, and I have zero reason to expect that Oil would support it in any way.

</details>

cc @grahamc @worldofpeace 

###### Known issues

~The current design of the Nix API assumes all of the shell files can be resolved with the same settings, and it won't be terribly obvious what to do (aside from break it up into multiple derivations, or override to manually invoke resholve) in these cases.~

~I'm not sure what the best approach here is. For reference, here's one possible approach from @grahamc: https://gsc.io/content-addressed/92d2781347e6e0d9b761ff2d4512d6885f48d8fb6a97db15f0ba5f4863ec91bd.nix~

I have pushed a ~tentative~ fix for this; see https://github.com/NixOS/nixpkgs/pull/85827#issuecomment-747831206 and d36bb7d

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions (in CI)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

abathur committed:
resholve: init

resholve attempts to resolve executables in shell scripts.
Includes Nix builder for resolving dependencies in Nix-built
shell projects.

abathur commented:
I marked this as draft for a few days while I refactored how it packages and consumes its Oil-shell dependency. Not super relevant, but the main changes are:

- It previously built from a fork of Oil where I fixed a few issues and added a setup.py, but I collapsed these changes into a set of patches maintained in the resholved repo.
- Oil isn't structured with being turned into a Python package in mind, so I also hacked through forcing an `oil` namespace on its subpackages, which were previously just leaking into the Python package namespace. Some of them have very generic names, like core, misc, frontend, tools, so I've been a little concerned about them causing namespace clashes. To make this work (without patching all of the oil source) resholved now rewrites oil's imports.

abathur commented on pkgs/development/misc/resholved/mk-resholved-derivation.nix:
```diff
@@ -0,0 +1,23 @@
+      inherit pname version src;
+      buildInputs = [ resholved ];
+      # tentatively disabled because gc probably knows things I don't :)
+      # propagatedBuildInputs = inputs;
```
I've tentatively disabled this after graham commented on IRC:

> gchristensen | hmm yeah I'm not sure why this is a thing:       propagatedBuildInputs = inputs;

I think he's right--this is vestigial--but I wanted to comment on it here until this is reviewed since this is something I wasn't sure how to handle in the first place.

---
Since the first working draft, resholved has followed source statements and resolved them along with the parent script. This behavior is important--it helps resholved weed functions from sourced libraries out of the list of commands it needs to resolve. The naive first version needed to propagate inputs to ensure they were available while resolving a consumer.

A later update added the tail "directive" comments that encode this resolve-time state for later reference, making it no longer necessary to propagate the inputs in order to resolve shell libraries.

---

While reasoning through this, I realized there's a related question about how to square shell norms and conventions with Nix norms and conventions. Some executables used in shell scripts/libraries are coincidental implementation details that a consumer probably shouldn't rely on (and their dependencies not propagated)--but sometimes they are integral and I'm a little less certain about whether they should come along?

Take, for example, two fairly popular bash dotfile managers: homeschick and yadm. Both of these programs are fairly thin wrappers around git, with varying degrees of reliance on the user also running bare git commands. I don't know what razor to use to decide if git should be propagated along with things like this?

emilazy replied:
I would err on the side of not propagating whenever it isn't required to make the software work. Explicit environment specification is a great thing, and propagation is a hack that should be avoided whenever possible; in an ideal world we wouldn't need it at all. (I can imagine cases where you'd want to deploy dotfiles from git without wanting to do actual commits and repository manipulation on that machine, for instance, though I admit this specific example isn't hugely convincing.)

abathur commented:
@grahamc I have finally released version 0.1.0 of resholve, updated this PR accordingly, and marked it ready for review.

abathur commented:
This doesn't include a maintainer entry for me, because I anticipated #84062 landing before this when I opened it. 

I've posted that one in prs-ready-for-review to try and nudge it along, but if it doesn't move I'll need to port my entry to this one.

abathur commented:
I'm OK with this merging without a solution to the question of resolving multiple scripts that need distinct rules/config, but I do want to try and solicit opinions on it.

Context: The current API enables you to specify a single set of dependencies/directives/configuration that apply to a single invocation of resholve (which can resholve one or more scripts). It is possible (not sure how common) for the needs of the scripts to conflict. For example, I noticed this while packaging one of my own projects with a profile script for sourcing (which should probably/ideally have no shebang, or a weird interpreter?), and one executable utility script (which should have a bash interpreter).

Here's an abbreviated example of how I have to handle it now: <a name="flat_example">
```nix
resholve.resholvePackage rec {
  # ...
  scripts = [ "bin/hag.bash" "bin/hag_import_history.bash" ];
  interpreter = "${bashInteractive_5}/bin/bash";
  # ...
}
```
I can go into more context if needed, but I'll cut to the chase for now: I've been wondering about modeling multiple runs as a set of sets, something like: <a name="nested_example">
```nix
resholve.resholvePackage rec {
  # ...
  jobs = { 
    profile_script = job { scripts = [ "bin/hag.bash" ]; interpreter = "none"; };
    util_cmds = job {
      scripts = ["bin/hag_import_history.bash" ];
      interpreter = "${bashInteractive_5}/bin/bash";
    };
  };
  #...
}
```
I'm not sure how hard it is to generate a script for setting envs and invoking resholve once per set from here, but I assume it is tractable.

I'm curious:
- is this an antipattern, or do you see other pitfalls to this approach?
- better ideas?
- better name than "job(s)"? 🤞

abathur committed:
resholve: 0.1.1 -> 0.2.0-1

I rewrote the Nix API to support multiple resholve invocations with
different settings.

abathur commented:
I've pushed a tentative (i.e., I'll release it if I don't get pushback here) update, `v0.2.0-1`, which reworks the Nix API so that it can specify more than one resholve invocation.

<a name="flat_example" /> Here's a (very abbreviated) example of how I was handling a package with one sourceable profile script, and one executable utility script: 
```nix
resholve.resholvePackage rec {
  # ...
  scripts = [ "bin/hag.bash" "bin/hag_import_history.bash" ];
  interpreter = "${bashInteractive_5}/bin/bash";
  # ...
}
```
<a name="nested_example" /> Here's how it looks with the API update: 
```nix
resholve.resholvePackage rec {
  # ...
  solutions = {
    profile = {
      scripts = [ "bin/hag.bash" ];
      interpreter = "none";
    };
    # profile/cmd are arbitrary names; could be a list
    # but I gather set is preferred for overridability
    cmd = {
      scripts = [ "bin/hag_import_history.bash" ];
      interpreter = "${bashInteractive_5}/bin/bash";
    };
  };
  #...
}
```

abathur committed:
resholve: 0.2.0-1 -> 0.2.0

abathur committed:
resholve: 0.2.0 -> 0.2.1

thiagokokada commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,150 @@
+  };
+
+  # resholve's primary dependency is this developer build of the oil shell.
+  oildev = python27.pkgs.buildPythonPackage rec {
```
You should use declare `buildPythonPackage` as a parameter instead, even if this package is only build using Python 2.7. To make this build fails with Python 3, use something like this: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/venusian/default.nix#L12 (but with `disabled = !isPy27` instead).

thiagokokada replied:
BTW, all those dependencies should be separate packages inside `pkgs/development/python-modules`, and the main package should use them as `buildDependencies`.

thiagokokada replied:
BTW, having oilshell and maybe oildev available in `all-packages` is probably a good idea, since those packages are kinda interesting.

thiagokokada replied:
You could, for example, open a PR with only `oildshell`, `oildev` and maybe `py-yajl`, making it easier to review only those dependencies. Afterwards you could make this PR depends on those dependencies.

abathur replied:
> BTW, all those dependencies should be separate packages inside `pkgs/development/python-modules`, and the main package should use them as `buildDependencies`. BTW, having oilshell and maybe oildev available in `all-packages` is probably a good idea, since those packages are kinda interesting. You could, for example, open a PR with only `oildshell`, `oildev` and maybe `py-yajl`, making it easier to review only those dependencies. Afterwards you could make this PR depends on those dependencies.

I have reservations about this. 

I'm happy to break these out if there's a clear imperative, but they are all pretty idiomatic (and all ultimately subservient to the oildev package itself). I don't anticipate them adding standalone value to nixpkgs in the short run.

1. The Oil shell already has a binary user package included here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/shells/oil/default.nix
2. oildev isn't (for now) a _real thing_ that the Oil project intentionally publishes--it's a fiction created by forcibly patching-and-packaging Oil's python code in order to build resholve
    - Because these Python APIs aren't intentionally published for external consumption and Oil is still a fast-changing young project, they aren't stable. I'd be very surprised if more than one consumer could share a single version of it for long.
    - AFAIK resholve is the only consumer of this. No one's asked me about using it.
    - resholve still needs [~140 lines of code](https://github.com/abathur/resholve/blob/d990fc06b504d761ff811bd1ad8075bdceb61dd0/resholve#L23-L161) just to monkeypatch all of oil's packages into a unified oil.* namespace. If anyone used it without this shim, all of its subpackages would leak into the root Python package namespace.
    - I can't guarantee any Oil code paths that resholve doesn't exercise actually work.
    - Because resholve is very tightly coupled to Oil's internal Python APIs, there's a fair chance any update of this would break it.
3. [py-yajl](https://github.com/oilshell/py-yajl) is Oil's fork of https://github.com/rtyler/py-yajl. They've modified it, dropped tests they don't need, and cut code they don't use. I have no idea how suitable it is or isn't for general use. Oil has made no attempt to promote their py-yajl fork, it has no stars, and I have zero reason to expect that Oil would support it in any way.

thiagokokada replied:
> The Oil shell already has a binary user package included here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/shells/oil/default.nix

I did a search for `oilshell` and didn't found anything, TIL.

thiagokokada commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,150 @@
+      ''; # breakers: doc, pgen2, vendor
+    };
+
+    # src = stdenv.lib.cleanSourceWith {
+    #   src = fetchFromGitHub {
+    #     owner = "oilshell";
+    #     repo = "oil";
+    #     rev = "ea80cdad7ae1152a25bd2a30b87fe3c2ad32394a";
+    #     sha256 = "0pxn0f8qbdman4gppx93zwml7s5byqfw560n079v68qjgzh2brq2";
+    #   };
+    #   filter = stdenv.lib.cleanSourceFilter;
+    # };
+
+    # src = stdenv.lib.sourceByRegex (fetchFromGitHub {
+    #   owner = "oilshell";
+    #   repo = "oil";
+    #   rev = "ea80cdad7ae1152a25bd2a30b87fe3c2ad32394a";
+    #   sha256 = "0pxn0f8qbdman4gppx93zwml7s5byqfw560n079v68qjgzh2brq2";
+    # }) [".*"];
```
Why all this commented code?

abathur replied:
```
/*
      This feels a little dumb; I assume it's wrongish. I was hoping
      to be able to filter the source down to knock out parts of the
      source that I'm not using, but I get errors just trying to get
      *anything* working with the approaches commented out below.

      On IRC eadwu[m] suggested dropping them in extraPostFetch.

      It's not critical to drop most of these; the primary target is
      the vendored fork of Python-2.7.13, which is ~ 55M and over 3200
      files, dozens of which get interpreter script patches in fixup.
      */
```

thiagokokada replied:
Well, I understand the issue, but I still won't sure that having commented code really helps.

Maybe adding just a note that you tried to use `stdenv.lib.cleanSourceWith` and `stdenv.lib.sourceByRegex` in this bigger comment and it didn't work?

abathur replied:
The comments can/should/will be removed before merge if someone says the way I'm doing it is fine (or points out how to do it correctly).

thiagokokada replied:
Ok, fair enough. It is just that I don't know if your PR is ready to review and merge or not.

It seems like you're looking more at feedback about it and less about trying to merge as is, did I understand it right?

thiagokokada replied:
Because if it is, probably my review isn't help here.

abathur replied:
Sorry--I believe it is roughly ready, but it is my first full package and it's covering some complexish ground where I had no clue if my decisions were appropriate. 

You are someone! If you say filtering them with `extraPostFetch` is fine by you, I'll cut the comments. If you aren't sure, I'll find someone to help clarify it.

thiagokokada replied:
For me it is fine, don't see the issue really.

thiagokokada replied:
BTW, I really would like if you could remove other comments from your code this would help. I think those comments makes the review harder.

You could for example convert those commented codes in inline comments like we are doing here.

abathur replied:
I pared comments back in 240bf06--is that what you had in mind?

thiagokokada commented on pkgs/development/misc/resholve/resholve-package.nix:
```diff
@@ -0,0 +1,107 @@
+    env:
+    value:
+    lib.escapeShellArg (makeEnvVal solution env value);
+  makeEnv = solution: env: value:
+    "RESHOLVE_${lib.toUpper env}=${shellEnv solution env value}";
+  scrub =
+    value:
+    removeAttrs value [ "scripts" "flags" ];
+  validateSolution = { scripts, inputs, interpreter, ... }: true;
+  makeEnvs =
+    solution:
+    value:
+    lib.concatStringsSep " " (lib.mapAttrsToList (makeEnv solution) (scrub value));
+  makeArgs =
+    { flags ? [ ], scripts, ... }:
+    lib.concatStringsSep " " (flags ++ scripts);
+  makeInvocation =
+    solution:
+    value:
+    if validateSolution value then
+      "${makeEnvs solution value} resholve --overwrite ${makeArgs value}"
+    else throw "invalid solution"; # shouldn't trigger for now
+  makeCommands =
+    solutions:
+    lib.mapAttrsToList makeInvocation solutions;
+
+  self = (stdenv.mkDerivation ((removeAttrs attrs [ "solutions" ])
+    // {
+    inherit pname version src;
+    buildInputs = [ resholve ];
+
+    #LOGLEVEL="INFO";
+    preFixup = ''
+      pushd $out
+      ${lib.concatStringsSep "\n" (makeCommands solutions)}
+      popd
+    '';
+  }));
+in
+lib.extendDerivation true passthru self
```
What this file is trying to do :thinking: ?

abathur replied:
It implements the resholvePackage function (i.e., as used in https://github.com/NixOS/nixpkgs/pull/107182/files)

thiagokokada reviewed (commented):
Squash your commits in only one with text `resholve: init at 0.2.1`.

thiagokokada commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,150 @@
+
+  py-yajl = python27.pkgs.buildPythonPackage rec {
+    pname = "oil-pyyajl";
+    version = "unreleased";
```
Add a date for the commit you're referring to.

abathur replied:
done in b0d38949a05

abathur commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,150 @@
+    nativeBuildInputs = [ re2c file ];
+
+    # runtime deps
+    propagatedBuildInputs = with python27.pkgs; [ python27 six typing ];
```
@thiagokokada Does the comment above at https://github.com/NixOS/nixpkgs/pull/85827#discussion_r550364088 apply here as well?

thiagokokada replied:
AFAIK, considering the complexity of this package you probably want to inject three parameters in the function input:

- `buildPythonApplication`
- `buildPythonPackage`
- `python27Packages`

And use them accordingly (in this case you would use `with python27Packages; [ six typing ]`. Also, I am not sure why you need to propagate `python27` here.

thiagokokada replied:
I may be wrong though, maybe someone with more Python package experience can help here.

abathur replied:
I think you're right on not propagating. Tests are passing local without it, at least. I'll give it a try in CI.

abathur replied:
I took a swing at this in b0d38949a05; does that look right, or did you intend for me to pass the buildPython functions in from all-packages? (I guess buildPython* are only passed by default for python-modules?)

thiagokokada replied:
Actually, this way looks mostly good. I just don't know what is better for those packages that will be stuck with Python 2.7 in future, if it is to use `pythonPackages` or `python27Packages`. I think the later is preferred because eventually we will switch to `pythonPackages` =  `python3Packages`, but I may be wrong.

thiagokokada replied:
Of course, you can also just leave as is and do `pythonPackages = python27Packages` in your `all-packages.nix` declaration.

SuperSandro2000 replied:
I would just use ``python27Packages`` as an input because you can't overwrite it right now anyway.

abathur replied:
@SuperSandro2000 Do you know if/how that should affect `disabled = !python27Packages.isPy27;` (ex: https://github.com/NixOS/nixpkgs/pull/85827/files#diff-f19dcdf25b44e15017ea999a1f918bb8f0485bf41e700a6b557e5b7c5c319669R41)?

abathur replied:
I updated these to python27Packages in 5bbea01. I'm not sure about `disabled = !python27Packages.isPy27;`, but it worked, so I'll tentatively resolve this.

thiagokokada replied:
Yeah, not problem in maintaining the `disabled = !python27Packages.isPy27`, makes it clear that the package is never meant to be built with Python 3.

abathur committed:
resholve: 0.2.1 -> 0.3.0

abathur committed:
use pythonPackages instead of python27.pkgs

abathur committed:
pare back comments

abathur commented:
> Squash your commits in only one with text `resholve: init at 0.2.1`.

Just wanted to acknowledge that I will squash to a single commit before merge.

lheckemann commented:
Ideally, we'd merge this using the squash and merge button, that way the history and context for the review comments is still visible in here.

thiagokokada commented:
Result of `nixpkgs-review pr 85827` [1](https://github.com/Mic92/nixpkgs-review)

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,79 @@
+    ./test.sh
+  '';
+
+  meta = {
+    description = "Resolve external shell-script dependencies";
+    homepage = "https://github.com/abathur/resholve";
+    license = with stdenv.lib.licenses; [
+      mit
+    ];
+    maintainers = with stdenv.lib.maintainers; [ abathur ];
+    platforms = stdenv.lib.platforms.all;
+  };
```
```suggestion
  meta = with stdenv.lib; {
    description = "Resolve external shell-script dependencies";
    homepage = "https://github.com/abathur/resholve";
    license = with licenses; [ mit ];
    maintainers = with maintainers; [ abathur ];
    platforms = platforms.all;
  };
```

abathur committed:
Update meta

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur commented:
> Result of `nixpkgs-review pr 85827` [1](https://github.com/Mic92/nixpkgs-review)
> 
> Seems that no result was generated by your PR 🤔 .

Not sure--I'm not familiar with nixpkgs-review yet. Perhaps exporting a set (`resholve.resholve`, and `resholve.resholvePackage`) is confusing it?

abathur committed:
use python27Packages instead of pythonPackages

SuperSandro2000 commented:
@ofborg eval

SuperSandro2000 commented:
> > Seems that no result was generated by your PR 🤔 .
> 
> Not sure--I'm not familiar with nixpkgs-review yet. Perhaps the exported set (`resholve.resholve`, and `resholve.resholvePackage`) is confusing it?

ofborg did not detect a changing packages.

lheckemann commented:
@ofborg build resholve.resholvePackage

@abathur we probably need `recurseIntoAttrs` on the resholve set for it to be detected by `nix-env -qa` and thus borg and nixpkgs-review, as well as hydra.

abathur committed:
use recurseIntoAttrs

abathur commented:
> @abathur we probably need `recurseIntoAttrs` on the resholve set for it to be detected by `nix-env -qa` and thus borg and nixpkgs-review, as well as hydra. Or put both resholve and resholvePackage at the top level.

@lheckemann Thanks--this seems to work. After committing https://github.com/NixOS/nixpkgs/pull/85827/commits/4098b2f08a39b510726fd3091370a527ddd1b25d I got the following out of nixpkgs-review:

```console
$ nix run nixpkgs#nixpkgs-review rev HEAD
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
$ git worktree add /Users/abathur/.cache/nixpkgs-review/rev-4098b2f08a39b510726fd3091370a527ddd1b25d/nixpkgs 0587d518db1b4c88eab4c09c902317a4dd9b2a7f
Preparing worktree (detached HEAD 0587d518db1)
Updating files: 100% (23689/23689), done.
HEAD is now at 0587d518db1 Merge pull request #107891 from lukegb/smokeping-prober
$ nix-env -f /Users/abathur/.cache/nixpkgs-review/rev-4098b2f08a39b510726fd3091370a527ddd1b25d/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit 4098b2f08a39b510726fd3091370a527ddd1b25d
Auto-merging pkgs/top-level/all-packages.nix
Automatic merge went well; stopped before committing as requested
$ nix-env -f /Users/abathur/.cache/nixpkgs-review/rev-4098b2f08a39b510726fd3091370a527ddd1b25d/nixpkgs -qaP --xml --out-path --show-trace --meta
1 package added:
resholve.resholve (init at 0.3.0)

$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /Users/abathur/.cache/nixpkgs-review/rev-4098b2f08a39b510726fd3091370a527ddd1b25d/build.nix
1 package built:
resholve.resholve

$ nix-shell /Users/abathur/.cache/nixpkgs-review/rev-4098b2f08a39b510726fd3091370a527ddd1b25d/shell.nix
```

thiagokokada reviewed (approved):
Code wise LGTM, I can also build with with `nixpkgs-review` but it conflicts with nixpkgs PATH so I can't post the result.

Anyway, approving, but I didn't test it very much (did run `resholve --help` and it works). Probably someone with more context of how the package is supposed to work can check the functionality.

Don't forget to squash the commits before merging.

abathur committed:
doc/comment on functions

abathur commented on pkgs/development/misc/resholve/resholve-package.nix:
```diff
@@ -0,0 +1,128 @@
+   *     flags = [ ];
+   *   };
+   * }
+   */
```
This doesn't seem like a terribly common idiom for similar functions. I don't expect this will actually stay in this file beyond the near term (I assume this will need to be explained at more length in the nixpkgs manual as it stabilizes), but I did want to leave some breadcrumbs.

abathur commented:
@SuperSandro2000 I haven't had any luck fishing for feedback on how to document this in the short run, so I pushed a commit updating the doc/comments where I previously had TODOs. 

That's the last thing I wanted to do here, so I think it is ready for a squash-and-merge?

SuperSandro2000 commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,124 @@
+      files, dozens of which get interpreter script patches in fixup.
+      */
+      extraPostFetch = ''
+        #find . -maxdepth 1 -type d | sort
+        rm -rf Python-2.7.13 benchmarks metrics py-yajl rfc gold web testdata services demo devtools cpp
+        # find . -maxdepth 1 -type d | sort
```
```suggestion
        rm -rf Python-2.7.13 benchmarks metrics py-yajl rfc gold web testdata services demo devtools cpp
```

Maybe move this into the commit if it is important for updates.

SuperSandro2000 commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,124 @@
+    # passing in from resholve.nix
+    patches = oilPatches;
+
+    buildInputs = [ readline cmark py-yajl makeWrapper ];
```
makeWrapper is a nativeBuildInputs.

abathur replied:
changed local

SuperSandro2000 commented on pkgs/development/misc/resholve/deps.nix:
```diff
@@ -0,0 +1,124 @@
+    propagatedBuildInputs = with python27Packages; [ six typing ];
+
+    doCheck = true;
+    dontStrip = true;
```
Does stripping break things?

abathur replied:
It doesn't look like it. Good catch; dropped in local copy.

SuperSandro2000 commented on pkgs/development/misc/resholve/default.nix:
```diff
@@ -0,0 +1,13 @@
+let
+  resholve = callPackage ./resholve.nix { inherit doCheck; };
+  resholvePackage =
+    callPackage ./resholve-package.nix { inherit resholve; };
```
Please move this into ``pkgs/build-support/resholve.nix``.

abathur replied:
Not keen to add extra friction to developing this out-of-tree in the resholve repo and porting fixes. 

I did go ahead and flatten the namespace out in https://github.com/NixOS/nixpkgs/pull/85827/commits/e9868c1ac77b9420b822a3df354023ffc3c47b7b, which I hope achieves most of your goal here.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6971,6 +6971,8 @@ in
 
   rescuetime = libsForQt5.callPackage ../applications/misc/rescuetime { };
 
+  resholve = recurseIntoAttrs (callPackage ../development/misc/resholve { });
```
Please call the package here with callPackage and move the build support like commented above.

abathur replied:
changed local

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,77 @@
+  checkInputs = [ bats ];
+  RESHOLVE_PATH = "${stdenv.lib.makeBinPath resolveTimeDeps}";
+
+  # explicit interpreter for test suite; maybe some better way...
+  INTERP = "${bash}/bin/bash";
+  checkPhase = ''
+    PATH=$out/bin:$PATH
```
```suggestion
  
  checkPhase = ''
    # explicit interpreter for test suite
    export INTERP="${bash}/bin/bash";
    export PATH=$out/bin:$PATH
```

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,77 @@
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    install resholve $out/bin/
```
```suggestion
    install -Dm755 resholve $out/bin/
```

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,77 @@
+  inherit version;
+  src = rSrc;
+  format = "other";
+  disabled = !python27Packages.isPy27;
```
```suggestion
```

Should be always true.

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,77 @@
+      "${rSrc}/0004-disable-internal-py-yajl-for-nix-built.patch"
+    ];
+  };
+  resolveTimeDeps = [ file findutils gettext ];
```
I think we can just move this down.

```suggestion
```

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve.nix:
```diff
@@ -0,0 +1,77 @@
+  '';
+  inherit doCheck;
+  checkInputs = [ bats ];
+  RESHOLVE_PATH = "${stdenv.lib.makeBinPath resolveTimeDeps}";
```
```suggestion
  RESHOLVE_PATH = "${stdenv.lib.makeBinPath [ file findutils gettext ]}";
```

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve-package.nix:
```diff
@@ -0,0 +1,128 @@
+    inherit pname version src;
+    buildInputs = [ resholve ];
+
+    #LOGLEVEL="INFO";
```
```suggestion
```

abathur replied:
Intentional breadcrumb to make it obvious how Nix user can get extra debug context out of the piñata. I've added a comment making this clearer local.

SuperSandro2000 commented on pkgs/development/misc/resholve/resholve-package.nix:
```diff
@@ -0,0 +1,128 @@
+    if builtins.isInt val then builtins.toString val
+    else if builtins.isString val then name
+    else if true == val then name
+    else if false == val then "" # omit!
+    else if null == val then "" # omit!
```
```suggestion
    else if false == val or null == val then "" # omit!
```

Not sure if this works in nix right now.

abathur replied:
It does. Ex: https://github.com/NixOS/nixpkgs/blob/487141bc9328a24d96e916a6ca71dffdf4a4c743/lib/systems/default.nix#L33-L45

SuperSandro2000 reviewed (commented):
Package should be fine after my suggestions. I am not to sure if the build support could be made easier or we have inbuild functions for some of the ones but it should be fine for now.

abathur committed:
cut commented lines

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
revert previous review addition

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
adopt review suggestion

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
revert previous review suggestion

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
brevity is the soul of wit

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
de-materialize resolveTimeDeps

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

abathur committed:
drop unnecessary dontStrip

abathur committed:
move makeWrapper to nativeBuildInputs

abathur committed:
explain commented loglevel

abathur committed:
fixup issues with review suggestions

abathur committed:
flatten namespace

abathur committed:
add resholve README

abathur committed:
resholve: 0.3.0 -> 0.4.0

abathur commented:
@SuperSandro2000 I've pushed commits that:
- mostly address the review comments here
- add a README.md as informal documentation, and a step towards eventually documenting it in the nixpkgs manual
- bump/release resholve 0.4.0 to track the breaking namespace change (the Nix API is included there as well; no other functionality change over the 0.3.0)

SuperSandro2000 commented:
Didn't read the entire doc but if we find mistake/areas to improve we can always do that later.

grahamc reviewed (approved):
I am exceedingly excited to use this!

worldofpeace commented:
It's a miracle :sparkles: @abathur I'm excited that you got this into nixpkgs :heart:
